File: bindings/typescript/src/triggers.ts
/**
 * The local trigger admin views (D113 / D170.b). A trigger binds an inbound EVENT
 * (a webhook POST, a cron interval, or a bare `SubmitTrigger` RPC) to a recipe
 * handle the event Invokes. The minimal-local single-user trigger; the hosted
 * multi-tenant trigger gateway at scale is CLOUD (GR19). Kept in its own module
 * (the `alerts.ts`/`toolscout.ts` module-per-concern precedent).
 *
 * SN-8: `triggerId`/`instanceId` are server-derived (the SDK only *encodes* the
 * bytes to hex); the auth secret is referenced by NAME only (never the value, a
 * `ListTriggers` row carries `authSecretPresent`, never the secret itself).
 */

import {
  type TriggerView as PbTriggerView,
  TriggerAuth,
  TriggerKind,
} from "./gen/kortecx/v1/gateway_pb.js";
import { encode } from "./hexids.js";

/** A registrable trigger kind (the kinds a client may create). */
export type TriggerKindInput = "webhook" | "cron" | "grpc";

/** A trigger kind as surfaced on a read. `"unknown"` absorbs UNSPECIFIED(0) + any
 *  future wire value (forward-compatible display). */
export type TriggerKindName = TriggerKindInput | "unknown";

/** A webhook auth posture a client may set. */
export type TriggerAuthInput = "none" | "hmac_sha256" | "bearer";

/** A trigger auth posture as surfaced on a read. `"unknown"` absorbs UNSPECIFIED(0)
 *  + any future wire value (forward-compatible display). */
export type TriggerAuthName = TriggerAuthInput | "unknown";

/** Map a friendly kind to the proto enum (closed input set). */
export function triggerKindToProto(kind: TriggerKindInput): TriggerKind {
  switch (kind) {
    case "webhook":
      return TriggerKind.WEBHOOK;
    case "cron":
      return TriggerKind.CRON;
    case "grpc":
      return TriggerKind.GRPC;
  }
}

/** Map a `TriggerKind` discriminant to a stable name (`"unknown"` if new). */
export function triggerKindName(kind: number): TriggerKindName {
  if (kind === TriggerKind.WEBHOOK) return "webhook";
  if (kind === TriggerKind.CRON) return "cron";
  if (kind === TriggerKind.GRPC) return "grpc";
  return "unknown";
}

/** Map a friendly auth posture to the proto enum (closed input set). */
export function triggerAuthToProto(auth: TriggerAuthInput): TriggerAuth {
  switch (auth) {
    case "none":
      return TriggerAuth.NONE;
    case "hmac_sha256":
      return TriggerAuth.HMAC_SHA256;
    case "bearer":
      return TriggerAuth.BEARER;
  }
}

/** Map a `TriggerAuth` discriminant to a stable name (`"unknown"` if new). */
export function triggerAuthName(auth: number): TriggerAuthName {
  if (auth === TriggerAuth.NONE) return "none";
  if (auth === TriggerAuth.HMAC_SHA256) return "hmac_sha256";
  if (auth === TriggerAuth.BEARER) return "bearer";
  return "unknown";
}

/** A `RegisterTrigger` request shape. The auth secret is referenced by NAME only
 *  (never the value, D81). The server derives the trigger id (SN-8). */
export interface RegisterTriggerInput {
  /** Unique operator handle (derives the trigger id). */
  readonly name: string;
  /** `"webhook"` | `"cron"` | `"grpc"`. */
  readonly kind: TriggerKindInput;
  /** The `kx/recipes/...` handle the event Invokes (`""` ⇒ App target). Exactly one
   *  of `recipeHandle` / `appHandle` is required. */
  readonly recipeHandle?: string;
  /** T-APP-TRIGGER-TARGET: a saved App handle the event runs via `RunApp` (`""` ⇒
   *  recipe target). The credentialed App fires unattended with its connections +
   *  secret_scope resolved. */
  readonly appHandle?: string;
  /** Webhook auth posture (defaults to `"none"`). */
  readonly auth?: TriggerAuthInput;
  /** SecretRef NAME of the HMAC/bearer secret (never the value; defaults to none). */
  readonly authSecretRef?: string;
  /** cron: interval seconds (`"300"`) OR a 5-field crontab expr (`"0 9 * * 1-5"`). */
  readonly scheduleSpec?: string;
  /** IANA timezone for a 5-field cron expr (e.g. `"America/New_York"`); `""` ⇒ UTC. */
  readonly timezone?: string;
  /** Defaults to `true`. */
  readonly enabled?: boolean;
  /** Per-trigger HITL (D114): withhold irreversible actions until an operator grant. */
  readonly requireApproval?: boolean;
}

/** The outcome of `registerTrigger` — the server-derived trigger id (hex). */
export interface RegisterTriggerResult {
  /** 16-byte server-derived trigger id, as lowercase hex. */
  readonly triggerId: string;
}

/** One governance row (`ListTriggers`). Never a secret value — `authSecretPresent`
 *  reports only whether a ref NAME is attached. */
export class TriggerRow {
  constructor(
    /** Server-derived trigger id, as lowercase hex. */
    readonly triggerId: string,
    readonly name: string,
    /** `"webhook"` | `"cron"` | `"grpc"` (`"unknown"` absorbs any future value). */
    readonly kind: TriggerKindName,
    /** `""` for an App target. */
    readonly recipeHandle: string,
    /** T-APP-TRIGGER-TARGET: the App target (`""` for a recipe target). */
    readonly appHandle: string,
    /** `"none"` | `"hmac_sha256"` | `"bearer"` (`"unknown"` absorbs any future value). */
    readonly auth: TriggerAuthName,
    /** A ref NAME is attached (never the value). */
    readonly authSecretPresent: boolean,
    readonly scheduleSpec: string,
    /** IANA timezone for a 5-field cron expr (`""` ⇒ UTC). */
    readonly timezone: string,
    readonly enabled: boolean,
    /** Per-trigger HITL posture (D114). */
    readonly requireApproval: boolean,
    /** Audit-only wall-clock; `0` ⇒ never fired. */
    readonly lastFireUnixMs: number,
  ) {}

  static fromProto(t: PbTriggerView): TriggerRow {
    return new TriggerRow(
      encode(t.triggerId),
      t.name,
      triggerKindName(t.kind),
      t.recipeHandle,
      t.appHandle,
      triggerAuthName(t.auth),
      t.authSecretPresent,
      t.scheduleSpec,
      t.timezone,
      t.enabled,
      t.requireApproval,
      Number(t.lastFireUnixMs),
    );
  }

  /** A plain snake_case object (stable wire-shaped serialization for UIs/logs). */
  toJSON() {
    return {
      trigger_id: this.triggerId,
      name: this.name,
      kind: this.kind,
      recipe_handle: this.recipeHandle,
      app_handle: this.appHandle,
      auth: this.auth,
      auth_secret_present: this.authSecretPresent,
      schedule_spec: this.scheduleSpec,
      timezone: this.timezone,
      enabled: this.enabled,
      require_approval: this.requireApproval,
      last_fire_unix_ms: this.lastFireUnixMs,
    };
  }
}

/** One `ListTriggers` page (deterministic `(name)` order). */
export interface TriggersPage {
  readonly triggers: readonly TriggerRow[];
  readonly hasMore: boolean;
}

/** The outcome of `submitTrigger` (the inbound EVENT verb) — the registered run
 *  (the PRIOR run when deduped) + whether a prior identical event already fired. */
export interface SubmitTriggerResult {
  /** 16-byte server-derived run instance id, as lowercase hex. */
  readonly instanceId: string;
  /** `true` ⇒ a prior identical event already started this run. */
  readonly deduped: boolean;
}

/** The outcome of `testTrigger` (a dry-run binding validation, fires nothing). */
export interface TestTriggerResult {
  readonly ok: boolean;
  /** Display-only validation prose (empty on success). */
  readonly detail: string;
}

File: bindings/python/src/kortecx/triggers.py
"""Trigger admin (D170 / D113) — durable webhook / cron / gRPC triggers that bind
an inbound event to a published recipe, surfaced by ``RegisterTrigger`` /
``ListTriggers`` / ``DeregisterTrigger`` / ``SubmitTrigger`` / ``TestTrigger``.

Kept in its own module (the feedback.py / module-per-concern precedent, GR3). The
``trigger_id`` + the bound ``instance_id`` are server-derived (the SDK only
hex-encodes the bytes); the kind / auth enums map to/from stable lowercase strings
(the ``rating_to_proto`` precedent — an unknown string is rejected fail-closed
with a ``ValueError``). The auth SECRET value is never on the wire — only whether
a ref NAME is attached (``auth_secret_present``, D81).
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import List

from . import hexids
from .v1 import gateway_pb2 as _g

# --- kind / auth enum <-> stable lowercase string -------------------------------

_KIND_TO_PROTO: "dict[str, _g.TriggerKind]" = {
    "webhook": _g.TriggerKind.WEBHOOK,
    "cron": _g.TriggerKind.CRON,
    "grpc": _g.TriggerKind.GRPC,
}
_KIND_NAMES: "dict[int, str]" = {v: k for k, v in _KIND_TO_PROTO.items()}

_AUTH_TO_PROTO: "dict[str, _g.TriggerAuth]" = {
    "none": _g.TriggerAuth.NONE,
    "hmac_sha256": _g.TriggerAuth.HMAC_SHA256,
    "bearer": _g.TriggerAuth.BEARER,
}
_AUTH_NAMES: "dict[int, str]" = {v: k for k, v in _AUTH_TO_PROTO.items()}


def trigger_kind_to_proto(kind: str) -> "_g.TriggerKind":
    """Map a friendly kind string (``"webhook"`` | ``"cron"`` | ``"grpc"``) to the
    proto ``TriggerKind`` enum. An unknown string is rejected fail-closed."""
    try:
        return _KIND_TO_PROTO[kind]
    except KeyError:
        raise ValueError(f"kind must be one of {sorted(_KIND_TO_PROTO)}, got {kind!r}") from None


def trigger_auth_to_proto(auth: str) -> "_g.TriggerAuth":
    """Map a friendly auth string (``"none"`` | ``"hmac_sha256"`` | ``"bearer"``)
    to the proto ``TriggerAuth`` enum. An unknown string is rejected fail-closed."""
    try:
        return _AUTH_TO_PROTO[auth]
    except KeyError:
        raise ValueError(f"auth must be one of {sorted(_AUTH_TO_PROTO)}, got {auth!r}") from None


def trigger_kind_name(kind: int) -> str:
    """Map a ``TriggerKind`` discriminant back to its stable lowercase name.
    ``"unknown"`` absorbs UNSPECIFIED(0) + any value this SDK predates (the
    ``lower_verdict_name`` precedent — never a crash)."""
    return _KIND_NAMES.get(kind, "unknown")


def trigger_auth_name(auth: int) -> str:
    """Map a ``TriggerAuth`` discriminant back to its stable lowercase name.
    ``"unknown"`` absorbs UNSPECIFIED(0) + any value this SDK predates (the
    ``lower_verdict_name`` precedent — never a crash)."""
    return _AUTH_NAMES.get(auth, "unknown")


# --- views ----------------------------------------------------------------------


@dataclass(frozen=True)
class TriggerView:
    """One registered trigger (``ListTriggers``). ``trigger_id`` is server-derived
    (hex); ``kind`` / ``auth`` are stable lowercase names; the auth SECRET value is
    never on the wire — only ``auth_secret_present`` (D81). ``last_fire_unix_ms``
    is audit-only wall clock (``0`` = never fired)."""

    trigger_id: str  # server-derived id, hex
    name: str
    kind: str  # "webhook" | "cron" | "grpc" | "unknown"
    recipe_handle: str  # "" for an App target
    app_handle: str  # T-APP-TRIGGER-TARGET: the App target ("" for a recipe target)
    auth: str  # "none" | "hmac_sha256" | "bearer" | "unknown"
    auth_secret_present: bool
    schedule_spec: str  # interval seconds OR a 5-field crontab expr (cron kind) / "" otherwise
    timezone: str  # IANA zone for a 5-field cron expr ("" ⇒ UTC)
    enabled: bool
    require_approval: bool  # per-trigger HITL posture (D114)
    last_fire_unix_ms: int

    @classmethod
    def from_proto(cls, t: "_g.TriggerView") -> "TriggerView":
        return cls(
            trigger_id=hexids.encode(t.trigger_id),
            name=t.name,
            kind=trigger_kind_name(t.kind),
            recipe_handle=t.recipe_handle,
            app_handle=t.app_handle,
            auth=trigger_auth_name(t.auth),
            auth_secret_present=t.auth_secret_present,
            schedule_spec=t.schedule_spec,
            timezone=t.timezone,
            enabled=t.enabled,
            require_approval=t.require_approval,
            last_fire_unix_ms=t.last_fire_unix_ms,
        )


@dataclass(frozen=True)
class TriggersPage:
    """One ``ListTriggers`` page (deterministic ``(name)`` order)."""

    triggers: List[TriggerView]
    has_more: bool
